README.md
## imagefluency: Image Statistics Based on Processing Fluency


## Overview

**imagefluency** is an simple R package for image fluency scores. The
package allows to get scores for several basic aesthetic principles that
facilitate fluent cognitive processing of images.
    
The main functions are:

* `img_contrast()`  to get the visual contrast of an image.
* `img_complexity()`  to get the visual complexity of an image (equals
   1 minus image simplicity)
* `img_self_similarity()`  to get the visual self-similarity of an image
* `img_simplicity()`  function to get the visual simplicity of an image (equals
   1 minus image complexity).
* `img_symmetry()`  to get the vertical and horizontal symmetry of an
   image.
* `img_typicality()`  to get the visual typicality of a list of images relative
   to each other

Other helpful functions are:

* `img_read()`  wrapper function to read images into R using `read.bitmap()` from the
  [readbitmap](https://github.com/jefferis/readbitmap) package
* `rgb2gray()`  convert images from RGB into grayscale (might speed up computation)
* `run_imagefluency()`  to launch a Shiny app locally on your computer for an interactive demo of the
   main functions


The main author is [Stefan Mayer](http://github.com/stm/).

## Installation

You can install the current stable version from CRAN.
```r
install.packages("imagefluency")
```

To download the latest development version from Github use the `install_github` function of the `devtools` package.
```r
# install devtools if necessary
if (!require("devtools")) install.packages("devtools")
# install imagefluency from github
devtools::install_github('stm/imagefluency')
```
Use the following link to report bugs/issues: <https://github.com/stm/imagefluency/issues>

## Example usage

```r
# visual contrast
#
# example image file (from package): bike.jpg
bike_location <- system.file("example_images", "bike.jpg", package = "imagefluency")
# read image from file
bike <- img_read(bike_location)
# get contrast
img_contrast(bike)

# visual symmetry
#
# read image
rails <- img_read(system.file("example_images", "rails.jpg", package = "imagefluency"))
# get only vertical symmetry
img_symmetry(rails, horizontal = FALSE)
```

## Documentation

See the package [vignette](https://stm.github.io/imagefluency/articles/imagefluency.html) for a detailled introduction (or type `vignette("imagefluency", package = "imagefluency")` into the R console) and the [reference page](https://stm.github.io/imagefluency/reference/index.html) for details on each function.

## Citation

If you want to cite this package in a scientific journal or in any other
context, run the following code in your `R` console:

``` r
utils::citation(package = "imagefluency")
```
There is currently a publication in preparation corresponding this
package and the citation will be updated once it’s published.


## Dependencies
The `img_complexity` function relies on the packages [R.utils](https://cran.r-project.org/package=R.utils) and [magick](https://github.com/ropensci/magick). The `img_self_similarity` function relies on the packages [OpenImageR](https://github.com/mlampros/OpenImageR), [pracma](https://cran.r-project.org/package=pracma), and [quadprog](https://cran.r-project.org/package=quadprog). The `img_read` function relies on the [readbitmap](https://github.com/jefferis/readbitmap) package. The `run_imagefluency` shiny app depends on [shiny](https://github.com/rstudio/shiny).

## References

* Mayer, S. & Landwehr, J, R. (2018). Quantifying Visual Aesthetics
Based on Processing Fluency Theory: Four Algorithmic Measures for
Antecedents of Aesthetic Preferences. *Psychology of Aesthetics,
Creativity, and the Arts*, *12*(4), 399--431. 
doi: [10.1037/aca0000187](https://doi.org/10.1037/aca0000187)

* Mayer, S. & Landwehr, J. R. (2018). Objective measures of design
typicality. *Design Studies*, *54*, 146--161.
doi: [10.1016/j.destud.2017.09.004](https://doi.org/10.1016/j.destud.2017.09.004)


NEWS.md
# imagefluency

# imagefluency 0.2.3

* updated unit tests

# imagefluency 0.2.2

* under-the-hood-changes in preparation for upcoming R 4.0.0 release

# imagefluency 0.2.1

* new package vignette
* new package website
* some under the hood code improvements

# imagefluency 0.2.0

Major update
* new package name: imagefluency
* shorter, more intuitive function names: img_* instead of quantify_*
* now works directly on color images
* improved symmetry score (finds best mirror axis)
* different compression algorithms for complexity
* interactive shiny app (alpha)

# imagefluency 0.1.2

* exemplary images included in package as data

# imagefluency 0.1.1

* complete rewrite of complexity function based on `magick` package

# imagefluency 0.1.0

* first version on github


R/complexity.R
#' @include utils.R
NULL

# 2do: idea / test: image compression via PCA (for images that are not in an uncompressed format)?
# see http://www.aaronschlegel.com/image-compression-principal-component-analysis/
#
# alternative: entropy as measure

#' Image complexity
#'
#' \code{img_complexity} returns the complexity of an image via image
#' compression. Higher values indicate higher image complexity.
#'
#' @details The function returns the visual complexity of an image. Visual
#'   complexity is calculated as ratio between the compressed and uncompressed
#'   image file size. Preferably, the original image is an uncompressed image
#'   file.
#'
#'   The function takes the file path of an image file (or URL) or a pre-loaded
#'   image as input argument (\code{imgfile}) and returns the ratio of the
#'   compressed divided by the uncompressed image file size. Values can range
#'   between almost 0 (virtually completely compressed image, thus extremly
#'   simple image) and 1 (no compression possible, thus extremely complex
#'   image).
#'
#'   You can choose between different image compression algorithms. Currently
#'   implemented are \code{zip} with deflate compression (default), \code{jpg},
#'   \code{gif}, and \code{png}. See Mayer & Landwehr (2018) for a discussion of
#'   different image compression algorithms for measuring visual complexity.
#'
#'   As most compression algorithms do not depict horizontal and vertical
#'   redundancies equally, the function includes an optional \code{rotate}
#'   parameter (default: \code{FALSE}). Setting this parameter to \code{TRUE}
#'   has the following effects: first, the image is rotated by 90 degrees.
#'   Second, a compressed version of the rotated image is created. Finally,
#'   the overall compressed image's file size is computed as the minimum of
#'   the original image's file size and the file size of the rotated image.
#'
#'   As \code{R}'s built-in \code{bmp} device creates (a) indexed instead of
#'   True Color images and (b) creates files with different file sizes depending
#'   on the operating system, the function relies on the \code{magick} package
#'   to write (and read) images.
#'
#' @param imgfile Either a character string containing the path to the image
#'   file (or URL) or an an image in form of a matrix (grayscale image) or array
#'   (color image) of numeric values representing the pre-loaded image (e.g. by
#'   using \code{\link{img_read}()}).
#' @param algorithm Character string that specifies which image compression
#'   algorithm to use. Currently implemented are \code{zip} with deflate
#'   compression (default), \code{jpg}, \code{gif}, and \code{png}.
#' @param rotate logical. Should the compressed file size of the rotated image
#'   also be computed? (see details)
#'
#' @return a numeric value: the ratio of the compressed divided by the
#'   uncompressed image file size
#' @export
#'
#' @examples
#' # Example image with high complexity: trees
#' trees <- img_read(system.file("example_images", "trees.jpg", package = "imagefluency"))
#' #
#' # display image
#' grid::grid.raster(trees)
#' #
#' # get complexity
#' img_complexity(trees)
#'
#'
#' # Example image with low complexity: sky
#' sky <- img_read(system.file("example_images", "sky.jpg", package = "imagefluency"))
#' #
#' # display image
#' grid::grid.raster(sky)
#' #
#' # get complexity
#' img_complexity(sky)
#'
#' @references Donderi, D. C. (2006). Visual complexity: A Review.
#'   \emph{Psychological Bulletin}, \emph{132}, 73--97.
#'   doi:\href{https://doi.org/10.1037/0033-2909.132.1.73}{10.1037/0033-2909.132.1.73}
#'
#' @references Forsythe, A., Nadal, M., Sheehy, N., Cela-Conde, C. J., & Sawey,
#'   M. (2011). Predicting Beauty: Fractal Dimension and Visual Complexity in
#'   Art. \emph{British Journal of Psychology}, \emph{102}, 49--70.
#'   doi:\href{https://doi.org/10.1348/000712610X498958}{10.1348/000712610X498958}
#'
#' @references Mayer, S. & Landwehr, J, R. (2018). Quantifying Visual Aesthetics
#'   Based on Processing Fluency Theory: Four Algorithmic Measures for
#'   Antecedents of Aesthetic Preferences. \emph{Psychology of Aesthetics,
#'   Creativity, and the Arts}, \emph{12}(4), 399--431.
#'   doi:\href{https://doi.org/10.1037/aca0000187}{10.1037/aca0000187}
#'
#'
#' @seealso \code{\link{img_read}}, \code{\link{img_contrast}},
#'   \code{\link{img_self_similarity}}, \code{\link{img_simplicity}},
#'   \code{\link{img_symmetry}}, \code{\link{img_typicality}},
img_complexity <- function(imgfile, algorithm = "zip", rotate = FALSE){

  if (!(algorithm %in% c("zip", "jpg", "gif", "png"))) {
    stop("Unknown image compression algorithm requested. Use one of the following: zip, jpg, gif, png", call. = FALSE)
  }

  if (is.character(imgfile) & !is.array(imgfile)) {
    if (length(imgfile) == 1) {
      input <- "image_path"
    } else {
      stop("Multiple filenames. Function can only handle one image at a time.", call. = FALSE)
    }
  } else if (is.array(imgfile) & (is.numeric(imgfile) | is.integer(imgfile))) {
    input <- "image"
    # normalize image if necessary
    if (max(imgfile) > 1) {
      imgfile <- imgfile / 255
    }
  } else {
    stop("Wrong type of input: has to be a filename (character string) or an image (3-dimensional array of numeric or integer values)", call. = FALSE)
  }

  if (!requireNamespace("magick", quietly = TRUE)) stop("Package 'magick' not found but needed. Please install the package first.", call. = FALSE)

  img <- tryCatch(magick::image_read(imgfile), error = function(err) err)
  if (inherits(img, "error")) {
    if (input == "image_path") {
      stop(paste0("File not found or invalid path (could not resolve '", imgfile ,"')"), call. = FALSE)
    } else {
      stop(paste0("File not found or invalid path (could not resolve input)"), call. = FALSE)
    }
  } else {
    .compl(img, algorithm, rotate)
  }
}


#' .compl
#'
#' Returns the complexity of an image array / matrix or path.
#'
#' @param imgfile An array or matrix of numeric values or integer values, or the file path to the image.
#' @param algorithm character. Which compression algorithm to use.
#' @param rotate logical. Image rotation by 90 degrees?
#'
#' @return a numeric value (ratio compressed/uncompressed file size).
#' @keywords internal
.compl <- function(img, algorithm, rotate) {
  flname <- file.path(tempdir(), .rand_string())

  # imginfo <- magick::image_info(img)

  # isBMP <- grepl("^BMP", as.character(imginfo[1]))
  # if (!isBMP) warning("Input image might not be uncompressed. Interpret results with caution.", call. = FALSE)

  # write bmp image
  magick::image_write(img, path = paste0(flname, ".bmp"), format = "bmp")
  orig_size <- file.size(paste0(flname, ".bmp"))

  # compress file
  if (algorithm == "zip") {
    if (!requireNamespace("R.utils", quietly = TRUE)) stop("Package 'R.utils' not found but needed. Please install the package first.", call. = FALSE)
    R.utils::gzip(filename = paste0(flname, ".bmp"))
  } else if (algorithm == "jpg") {
    magick::image_write(img, path = paste0(flname, ".jpg"), format = "jpg")
  } else if (algorithm == "gif") {
    magick::image_write(img, path = paste0(flname, ".gif"), format = "gif")
  } else if (algorithm == "png") {
    magick::image_write(img, path = paste0(flname, ".png"), format = "png")
  }

  # read file sizes
  if (algorithm == "zip") compressed_size <- file.size(paste0(flname,".bmp.gz"))
  if (algorithm == "jpg") compressed_size <- file.size(paste0(flname,".jpg"))
  if (algorithm == "gif") compressed_size <- file.size(paste0(flname,".gif"))
  if (algorithm == "png") compressed_size <- file.size(paste0(flname,".png"))

  # remove newly created files
  # Sys.sleep(0.25) # wait a moment
  if (algorithm == "zip") file.remove(paste0(flname, ".bmp.gz"))
  if (algorithm == "jpg") file.remove(paste0(flname, ".jpg"))
  if (algorithm == "gif") file.remove(paste0(flname, ".gif"))
  if (algorithm == "png") file.remove(paste0(flname, ".png"))
  # remove original file (if not zip, where this is done automatically)
  if (algorithm != "zip")   file.remove(paste0(flname, ".bmp"))

  if (!is.logical(rotate)) {
    warning(paste0("rotate = '", rotate, "' is not a logical value (TRUE/FALSE). Skipping rotation ..."), call. = FALSE)
    rotate <- FALSE
  }

  if (rotate) {
    flname_rot <- paste0(flname, "_rot")

    # message("... trying to rotate the image ...")
    # write rotated image
    magick::image_write(magick::image_rotate(img, degrees = 90), path = paste0(flname_rot, ".bmp"), format = "bmp")

    # compress rotated image
    if (algorithm == "zip") {
      R.utils::gzip(filename = paste0(flname_rot, ".bmp"))
    } else if (algorithm == "jpg") {
      magick::image_write(magick::image_rotate(img, degrees = 90), path = paste0(flname_rot, ".jpg"), format = "jpg")
    } else if (algorithm == "gif") {
      magick::image_write(magick::image_rotate(img, degrees = 90), path = paste0(flname_rot, ".gif"), format = "gif")
    } else if (algorithm == "png") {
      magick::image_write(magick::image_rotate(img, degrees = 90), path = paste0(flname_rot, ".png"), format = "png")
    }

    # read file size
    if (algorithm == "zip") compressed_size_rot <- file.size(paste0(flname_rot,".bmp.gz"))
    if (algorithm == "jpg") compressed_size_rot <- file.size(paste0(flname_rot,".jpg"))
    if (algorithm == "gif") compressed_size_rot <- file.size(paste0(flname_rot,".gif"))
    if (algorithm == "png") compressed_size_rot <- file.size(paste0(flname_rot,".png"))

    # remove rotated files
    # Sys.sleep(0.25) # wait a moment
    if (algorithm == "zip") file.remove(paste0(flname_rot,".bmp.gz"))
    if (algorithm == "jpg") file.remove(paste0(flname_rot,".jpg"))
    if (algorithm == "gif") file.remove(paste0(flname_rot,".gif"))
    if (algorithm == "png") file.remove(paste0(flname_rot,".png"))
    # remove original file (if not zip, where this is done automatically)
    if (algorithm != "zip") file.remove(paste0(flname_rot, ".bmp"))

    # update result: minimum of both compressed file sizes of original and rotated img
    compressed_size <- min(compressed_size, compressed_size_rot)
  }

  # remove initial tempfile


  # return ratio between compressed and original (bmp) size (i.e., compression rate)
  return(compressed_size/orig_size)
}


# # not in use yet
# img_entropy <- function(img) {
#
#   binMin <- 0
#   binMax <- 1
#   if (max(img) > 1) binMax <- 255
#
#   # calculate histogram (with 256 bins as in Matlab)
#   imgHist <- hist(as.vector(img), breaks = seq(binMin, binMax, length.out = 256), plot = FALSE)$density
#   #imgHist <- hist(img, breaks = seq(binMin, binMax, length.out = 256), plot = FALSE)$counts
#
#   # remove zero entries
#   imgHist[imgHist==0] <- NA
#   imgHist <- na.omit(imgHist)
#
#   # normalize
#   imgNorm <- imgHist
#   #imgNorm <- imgHist / sum(imgHist)
#
#   # entropy
#   entr <- -sum(imgNorm * log2(imgNorm))
#   return(entr)
# }
